registration-processor/pre-processor/registration-processor-packet-validator-stage/src/main/java/io/mosip/registration/processor/stages/utils/XSDValidation.java

```diff
@@ -0,0 +1,99 @@
+package io.mosip.registration.processor.stages.utils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.io.IOUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import io.mosip.kernel.biometrics.constant.BiometricType;
+import io.mosip.kernel.biometrics.entities.BiometricRecord;
+import io.mosip.kernel.cbeffutil.container.impl.CbeffContainerImpl;
+import io.mosip.kernel.core.cbeffutil.common.CbeffValidator;
+import io.mosip.kernel.core.cbeffutil.entity.BDBInfo;
+import io.mosip.kernel.core.cbeffutil.entity.BIR;
+import io.mosip.kernel.core.cbeffutil.entity.BIRInfo;
+import io.mosip.kernel.core.cbeffutil.entity.BIRVersion;
+import io.mosip.kernel.core.cbeffutil.jaxbclasses.BIRType;
+import io.mosip.kernel.core.cbeffutil.jaxbclasses.ProcessedLevelType;
+import io.mosip.kernel.core.cbeffutil.jaxbclasses.PurposeType;
+import io.mosip.kernel.core.cbeffutil.jaxbclasses.QualityType;
+import io.mosip.kernel.core.cbeffutil.jaxbclasses.RegistryIDType;
+import io.mosip.kernel.core.cbeffutil.jaxbclasses.SingleType;
+@Component
+public class XSDValidation {
+
+    @Value("${mosip.kernel.xsdstorage-uri}")
+    private String configServerFileStorageURL;
+    
+    @Value("${mosip.kernel.xsdfile}")
+    private String schemaName;
+    
+    public boolean validateXSD(BiometricRecord biometricRecord ) throws IOException, Exception {
+    	try (InputStream xsd = new URL(configServerFileStorageURL + schemaName).openStream()) {
+            CbeffContainerImpl cbeffContainer = new CbeffContainerImpl();
+            List<BIR> birList = new ArrayList<>();
+            biometricRecord.getSegments().forEach(s -> birList.add(convertToBIR(s)));
+            BIRType bir = cbeffContainer.createBIRType(birList);
+            return CbeffValidator.createXMLBytes(bir, IOUtils.toByteArray(xsd))!=null;//validates XSD 

```

Comment:
While developing, did we print the XML this line has generated and compared it with the input cbeff xml file in the packet ?

Reply:
"BIRType birType = CbeffValidator.getBIRFromXML(IOUtils.toByteArray(biometrics));" is being used in packet manager to extract biometric record from packet. Following code is being used to extract data. 

public static BIRType getBIRFromXML(byte[] fileBytes) throws Exception {
		JAXBContext jaxbContext = JAXBContext.newInstance(BIRType.class);
		Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
		JAXBElement<BIRType> jaxBir = unmarshaller.unmarshal(new StreamSource(new ByteArrayInputStream(fileBytes)),
				BIRType.class);
		BIRType bir = jaxBir.getValue();
		return bir;
	}

If the data is of incorrect format we will get an unmarshalling exception.

Reply:
So the scenario is XML is proper (means proper xml structure) but it is not having attribute or tag that is maditory as per the XSD